zhaquirks/inovelli/__init__.py

```diff
@@ -0,0 +1,206 @@
+"""Module for Inovelli quirks implementations."""
+
+import logging
+from typing import Any, List, Optional, Union
+
+from zigpy.quirks import CustomCluster
+import zigpy.types as t
+from zigpy.zcl import foundation
+from zigpy.zcl.clusters.manufacturer_specific import ManufacturerSpecificCluster
+
+from zhaquirks.const import (
+    BUTTON,
+    BUTTON_1,
+    BUTTON_2,
+    BUTTON_3,
+    COMMAND,
+    COMMAND_DOUBLE,
+    COMMAND_HOLD,
+    COMMAND_ID,
+    COMMAND_PRESS,
+    COMMAND_QUAD,
+    COMMAND_RELEASE,
+    COMMAND_TRIPLE,
+    DOUBLE_PRESS,
+    PRESS_TYPE,
+    QUADRUPLE_PRESS,
+    QUINTUPLE_PRESS,
+    TRIPLE_PRESS,
+    ZHA_SEND_EVENT,
+)
+
+_LOGGER = logging.getLogger(__name__)
+INOVELLI_VZM31SN_CLUSTER_ID = 64561
+
+# Press Types
+# 0 - pressed
+# 1 - released
+# 2 - held
+# 3 - 2x
+# 4 - 3x
+# 5 - 4x
+# 6 - 5x
+COMMAND_QUINTUPLE = "quintuple"
+PRESS_TYPES = {
+    0: COMMAND_PRESS,
+    1: COMMAND_RELEASE,
+    2: COMMAND_HOLD,
+    3: COMMAND_DOUBLE,
+    4: COMMAND_TRIPLE,
+    5: COMMAND_QUAD,
+    6: COMMAND_QUINTUPLE,
+}
+
+# Buttons
+# 1 - down button
+# 2 - up button
+# 3 - config button
+
+BUTTONS = {1: BUTTON_1, 2: BUTTON_2, 3: BUTTON_3}
+ON = "Up"
+OFF = "Down"
+CONFIG = "Config"
+
+
+class Inovelli_VZM31SN_Cluster(CustomCluster):
+    """Inovelli VZM31-SN custom cluster."""
+
+    cluster_id = 0xFC31
+    name = "InovelliVZM31SNCluster"
+    ep_attribute = "inovelli_vzm31sn_cluster"
+
+    attributes = ManufacturerSpecificCluster.attributes.copy()
+    attributes.update(
+        {
+            0x0001: ("Dimming_Speed_Up_Remote", t.uint8_t),
```

Comment:
To ZCL attribute definitions, make them snake case and add a `, True` at the end of the definition tuple to indicate that the attribute is manufacturer-specific.

```Python
0x0001: ("dimming_speed_up_remote", t.uint8_t, True),
          ^       ^     ^  ^                 ^^^^^^
```

If you have more information about the access controls on each specific attribute, you can use the full attribute definition object instead of the shorthand tuple:

```Python
from zigpy.zcl import foundation

...
0x0001: foundation.ZCLAttributeDef("dimming_speed_up_remote", t.uint8_t, access="rw", is_manufacturer_specific=True)
```

Reply:
I should be able to tackle these along with the rest here soon fortunately. I'll have to test the manufacturer-specific piece as I don't think that worked previously for me though.

Out of curiosity, is there any alternative method that lets us have a more user-friendly name in the UI? Not sure if we'd need to go the full blueprint route for something more user-friendly, but these attributes will be used to configure the switch and edit parameters, etc. I'm new to messing with ZHA quirks so not sure if I just haven't figured it out yet, but it seems like I'm a bit limited in terms of information provided in Manage Clusters, is that the point of blueprints?

Reply:
The manage clusters UI is really like a developer mode. We are working on adding configuration entity support to ZHA.

Reply:
Got it, that makes sense and looking forward to it, thanks!

Reply:
> The manage clusters UI is really like a developer mode. We are working on adding configuration entity support to ZHA.

Oh heck yes. This will be huge since this device has like 60+ parameters. Z-wave JS obviously is a fantastic template to go off of...